crates/iceberg/src/metadata_scan.rs

```diff
@@ -0,0 +1,267 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Metadata table api.
+
+use std::sync::Arc;
+
+use arrow_array::builder::{MapBuilder, PrimitiveBuilder, StringBuilder};
+use arrow_array::types::{Int64Type, TimestampMillisecondType};
+use arrow_array::RecordBatch;
+use arrow_schema::{DataType, Field, Schema, TimeUnit};
+
+use crate::spec::TableMetadataRef;
+use crate::table::Table;
+use crate::Result;
+
+/// Table metadata scan.
+#[derive(Debug)]
+pub struct MetadataScan {
+    metadata_ref: TableMetadataRef,
+}
+
+impl MetadataScan {
+    /// Creates a new metadata scan.
+    pub fn new(table: &Table) -> Self {
+        Self {
+            metadata_ref: table.metadata_ref(),
+        }
+    }
+
+    /// Returns the snapshots of the table.
+    pub fn snapshots(&self) -> Result<RecordBatch> {
+        SnapshotsTable::scan(self)
+    }
+}
+
+/// References:
+/// - <https://github.com/apache/iceberg/blob/ac865e334e143dfd9e33011d8cf710b46d91f1e5/core/src/main/java/org/apache/iceberg/MetadataTableType.java#L23-L39>
+/// - <https://iceberg.apache.org/docs/latest/spark-queries/#querying-with-sql>
+/// - <https://py.iceberg.apache.org/api/#inspecting-tables>
+pub trait MetadataTable {
+    /// Returns the schema of the metadata table.
+    fn schema() -> Schema;
```

Comment:
I need to make this separate and public, so engines can get the schema first without fetching the data.

Reply:
Some of the metadata tables (e.g. `entires`) have schemas that depend on the table schema. E.g. `readable_metrics` is a struct with field per column. Might be worth having this take a `scan` as well or do so in another PR.